File: jbake-dist/src/smoke-test/java/org/jbake/BuiltInProjectsTest.java
package org.jbake;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.condition.OS;
import org.junit.jupiter.api.io.TempDir;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import static org.assertj.core.api.Assertions.assertThat;

class BuiltInProjectsTest {

    @TempDir
    private Path folder;
    private File projectFolder;
    private File templateFolder;
    private File outputFolder;
    private String jbakeExecutable;
    private BinaryRunner runner;

    @BeforeEach
    void setup() throws IOException {
        if (OS.current() == OS.WINDOWS) {
            jbakeExecutable = new File("build\\install\\jbake\\bin\\jbake.bat").getAbsolutePath();
        } else {
            jbakeExecutable = new File("build/install/jbake/bin/jbake").getAbsolutePath();
        }
        projectFolder = folder.resolve("project").toFile();
        Files.createDirectory(projectFolder.toPath());
        templateFolder = new File(projectFolder, "templates");
        outputFolder = new File(projectFolder, "output");
        runner = new BinaryRunner(projectFolder);
    }

    @ParameterizedTest
    @CsvSource({
        "thymeleaf,thyme",
        "freemarker,ftl",
        "jade,jade",
        "groovy,gsp",
        "groovy-mte,tpl"
    })
    void shouldBakeWithProject(String projectName, String extension) throws Exception {
        shouldInitProject(projectName, extension);
        shouldBakeProject();
    }

    private void shouldInitProject(String projectName, String extension) throws IOException, InterruptedException {
        Process process = runner.runWithArguments(jbakeExecutable, "-i", "-t", projectName);
        assertThat(process.exitValue()).isEqualTo(0);
        assertThat(new File(projectFolder, "jbake.properties")).exists();
        assertThat(new File(templateFolder, String.format("index.%s", extension))).exists();
        process.destroy();
    }

    private void shouldBakeProject() throws IOException, InterruptedException {
        Process process = runner.runWithArguments(jbakeExecutable, "-b");
        assertThat(process.exitValue()).isEqualTo(0);
        assertThat(new File(outputFolder, "index.html")).exists();
        process.destroy();
    }

}

File: jbake-dist/src/smoke-test/kotlin/org/jbake/BuiltInProjectsTest.kt
package org.jbake

import io.kotest.assertions.withClue
import io.kotest.core.spec.style.FunSpec
import io.kotest.datatest.withData
import io.kotest.matchers.file.shouldExist
import io.kotest.matchers.shouldBe
import kotlin.io.path.createTempDirectory

data class TemplateTestCase(val name: String, val extension: String) {
    override fun toString() = "$name template"
}

class BuiltInProjectsTest : FunSpec({

    context("should bake successfully") {
        withData(
            TemplateTestCase("freemarker", "ftl"),
            TemplateTestCase("thymeleaf", "thyme"),
            TemplateTestCase("jade", "jade"),
            TemplateTestCase("groovy", "gsp"),
            TemplateTestCase("groovy-mte", "tpl"),
        ) { (projectName, extension) ->
            val jbakeExec = BinaryRunner.jbakeExecutableRelative.absolutePath
            val tempDir = createTempDirectory(buildOutputDir, "jbake-smoke-test-").toFile()
            try {
                val runner = BinaryRunner(tempDir)

                // Run JBake to initialize the dir (with verbose logging to debug failures)
                val initProcess = runner.runWithArguments(jbakeExec, "-vvv", "-i", "-t", projectName)
                withClue("JBake init run printed:\n\n${runner.processOutput}\n") {
                    initProcess.exitValue() shouldBe 0
                }
                tempDir.resolve("jbake.properties").shouldExist()
                tempDir.resolve("templates/index.$extension").shouldExist()
                initProcess.destroy()

                // Bake project (with verbose logging to debug failures)
                val bakeProcess = runner.runWithArguments(jbakeExec, "-vvv", "-b")
                withClue("JBake bake run printed:\n\n${runner.processOutput}\n") {
                    bakeProcess.exitValue() shouldBe 0
                }
                tempDir.resolve("output/index.html").shouldExist()
                bakeProcess.destroy()
            } finally {
                tempDir.deleteRecursively()
            }
        }
    }
})
